lib/miq_cockpit.rb

```diff
@@ -0,0 +1,143 @@
+        else
+          url.scheme = "http"
+        end
+        url.port = opts[:port] || DEFAULT_PORT
+      end
+
+      if !url.path.ends_with?("/")
+        url.path = url.path + "/"
+      end
+
+      return url
+    end
+
+    def initialize(opts = {})
+      if !opts
+        opts = {}
+      end
+
+      @opts = opts
+      @config_dir = File.join(File.dirname(__FILE__), '../config')
+    end
+
+    def config_dir
+      @config_dir
+    end
+
+    def command (address)
+      cmd = ['/usr/libexec/cockpit-ws', "--port", @opts[:port] || "#{DEFAULT_PORT}"]
+      if (address)
+        cmd = cmd.push("--address", address)
+      end
+      if (!@opts[:ssl])
+        cmd = cmd.push("--no-tls")
+      end
+      return cmd
+    end
+
+    def save_config
+      Dir.mkdir(@config_dir) unless File.exists?(@config_dir)
+      Dir.mkdir(File.join(@config_dir, "cockpit")) unless File.exists?(File.join(@config_dir, "cockpit"))
+      fname = File.join(@config_dir, "cockpit", "cockpit.conf")
+      update()
+      File.open(fname, "w") { |f| f.write @config }
+      if @opts[:ssl]
+        setup_ssl()
+      end
+    end
+
+    def update
+      title = "ManageIQ Cockpit"
+
+      login_command =  File.join(File.dirname(__FILE__), 'cockpit', 'cockpit-auth-miq')
+      miq_url = @opts[:miq_url]
+      @config = <<-END_OF_CONFIG
+[Webservice]
+LoginTitle = #{title}
+UrlRoot = cws
+
+[Negotiate]
+Action = none
+
+[Basic]
+Action = none
+
+[Bearer]
+Action = remote-login-ssh
+
+[SSH-Login]
+command = #{login_command}
+authFD=10
+
+[OAuth]
+Url = #{miq_url}/dashboard/cockpit_redirect
+
+END_OF_CONFIG
+      @config
+    end
+
+    def setup_ssl
+      dir = File.join(@config_dir, "cockpit", "ws-certs.d")
+      Dir.mkdir(dir) unless File.exists?(dir)
+      cert = OpenSSL::X509::Certificate.new(File.open('certs/server.cer')).to_pem()
+      key = OpenSSL::PKey::RSA.new(File.open('certs/server.cer.key')).to_pem()
+      File.open(File.join(dir, "0-miq.cert"), "w") { |f|
+        f.write(cert)
+        f.write("\n")
+        f.write(key)
+      }
+    end
+  end
+
+  class ApacheConfig
+    def initialize(opts = {})
+      update(opts)
+    end
+
+    def save(fname)
+      File.open(fname, "w") { |f| f.write @config }
+    end
+
+    def update(opts)
+      url = MiqCockpit::WS.url(opts, "localhost")
+      ws_scheme = url.scheme == "https" ? "wss" : "ws"
+      ssl_status = url.scheme == "https" ? "on" : "off"
+      scheme = url.scheme
+      port = url.port
+      path = url.path
+      host = url.host
+
+      @config = <<-END_OF_CONFIG
+    <LocationMatch "^/cws/cockpi(t[^/]+|t)?/socket$">
+        SSLProxyEngine #{ssl_status}
+        SetEnvIf Origin "(http|https)://(.*)" origin=$2
+        RequestHeader set Origin "#{scheme}://#{host}:#{port}" "expr=%{HTTP_HOST} == %{env:origin}"
+        ProxyPassMatch "#{ws_scheme}://#{host}:#{port}#{path}cockpi$1/socket"
+    </LocationMatch>
+
+    ProxyPass /cws/ #{scheme}://#{host}:#{port}#{path}
```

Comment:
string concatention will definitely fail with ipv6 unless you manually wrap the address.

Use `URI::Generic.build(:scheme => scheme, :host => host, ...)` instead.